salt: Add a  `metalk8s:debug` pillar option

gdemonet opened the pull request:
**Component**: salt, kubernetes

**Context**:

When trying to troubleshoot strange issues, we are often missing detailed logs. Enabling such logs requires manual operations to be performed **after** the initial installation (hence, it's not possible to have said logs during the initial installation).

**Summary**:

With this PR, we introduce a (really crude) solution to the above problem, with a single `debug` toggle in BootstrapConfiguration. When enabled, all the following components will generate more logs:

- Salt master and minions
- containerd
- kubelet
- kube-apiserver, kube-controller-manager, kube-scheduler, kube-proxy
- CoreDNS

_**Note**: for a `single-node-install-rhel` stage, the logs collected on bootstrap go from around 10MB (uncompressed) without the `metalk8s:debug` pillar enabled, to 45MB when it's enabled_

**Acceptance criteria**: 

Keeps working :) there will probably be a performance impact of running everything like this, so for now, it's not documented, and shouldn't be used in production. We'll see where to go from there.

gdemonet committed:
ci: Add more logs to Eve upgrade/downgrade

We didn't use the `--verbose` option.

gdemonet committed:
salt: Add 'metalk8s:debug' pillar

In order to provide a simple toggle to users at the BootstrapConfig
level, we setup this `debug` option, which will be used across all
formulas to setup debug logging/sockets where applicable.
This is only a very basic approach for now (either everything is in
debug mode, or not), which should be improved on later.
However, it should provide some form of debugging help which we don't
have at the moment.

gdemonet committed:
salt: Use debug pillar in Salt master states

Note: we remove the --log-level argument from salt-master container, but
not from salt-api, as it doesn't expose a log-level configuration
option.

gdemonet committed:
salt: Use debug pillar in Salt minion states

Note that when `metalk8s.debug` is enabled, the various scripts
interacting with Salt minion will output logs at the [DEBUG] level.
This may be further refined by adding verbosity controls at our scripts
level (but will probably wait for the Python part of #2163).

gdemonet committed:
salt: Use debug pillar in containerd states

Note that we extend the systemd service drop-in to modify the
`ExecStart` directive, which allows us to add some command-line
arguments (if they were not exposed through environment vars).
We use this new flexibility to add a `--log-level`, and set it to
`debug` if `metalk8s:debug` pillar is enabled.

gdemonet committed:
salt: Use debug pillar in K8s control plane states

k-a, k-c-m and k-s all use the same logging framework, which can be
controlled using the `--v` verbosity argument. We set it to 2 for now
when `metalk8s:debug` is enabled.

gdemonet committed:
salt: Use debug pillar in coredns states

CoreDNS doesn't log by default, so we enable this only if
`metalk8s:debug` is enabled.

gdemonet committed:
salt: Use debug pillar in kube-proxy states

kube-proxy follows the same pattern as K8s control plane components,
using the `--v` option.

gdemonet committed:
salt: Use debug pillar in kubelet states

kubelet uses the same logging framework as other Kubernetes components,
hence uses the same strategy with the `--v` command-line arg (set to 0
by default, to 2 when in debug mode).

gdemonet committed:
ci/terraform: Add debug in BootstrapConfig

We add a `debug` variable to Terraform modules (in `single-node-rhel`
and `multiple-nodes-centos`), and use it to enable the `debug` option in
BootstrapConfiguration.

gdemonet committed:
ci: Add a 'metalk8s_debug' build property

This property sets the `debug` option in BootstrapConfiguration, for all
deployments: single/multiple node(s), RHEL/CentOS,
install/upgrade/downgrade...
Finer control may be added in the future, but requires a lot of
boilerplate in Eve's YAML to handle, so not right now.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/add-debug-pillar` with contents from `improvement/add-debug-pillar`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/add-debug-pillar origin/development/2.5
 $ git merge origin/improvement/add-debug-pillar
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/add-debug-pillar
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-debug-pillar` into
`development/2.4`
* [w/2.5/improvement/add-debug-pillar](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/add-debug-pillar) will be merged into `development/2.5` 
* [w/2.6/improvement/add-debug-pillar](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/add-debug-pillar) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

TeddyAndrieux commented on salt/metalk8s/salt/master/files/salt-master-manifest.yaml.j2:
```diff
@@ -82,7 +81,8 @@ spec:
       image: {{ image }}
       command: ['salt-api']
-      args: ['--log-level=info']
+      args:
+        - --log-level={{ 'debug' if debug else 'info' }}
```
Salt-api do not use the value from master config ?

gdemonet replied:
It's a good question! Not clearly stated in documentation, and I didn't test...

gdemonet replied:
It does :) Thanks for the tip!

bert-e commented:
# History mismatch



Merge commit #74d898a1c1a1d91b80e526506ac9d64373f154bd on the integration branch
`w/2.5/improvement/add-debug-pillar` is merging a branch which is neither the current
branch `improvement/add-debug-pillar` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/add-debug-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
ci: Collect logs on failure

Some steps were not always run, hence we were missing sosreport archives
after a failed install / test.

bert-e commented:
# History mismatch



Merge commit #5f97664e1e85e3f507daf00cf83f7f811624eff7 on the integration branch
`w/2.5/improvement/add-debug-pillar` is merging a branch which is neither the current
branch `improvement/add-debug-pillar` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/add-debug-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented on salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2:
```diff
@@ -1,4 +1,7 @@
 [Service]
+# Allow passing arguments to containerd daemon
+ExecStart=
+ExecStart=/usr/bin/containerd {{ args | join(' ') }}
```
Not really resilient if there is space in one of the args... but I assume it'll not be the case, at least it's not for now.

gdemonet replied:
I can trick it like this:

```jinja
/usr/bin/containerd "{{ args | join('" "'}}"
```

gdemonet replied:
Done

alexandre-allard commented on salt/metalk8s/container-engine/containerd/installed.sls:
```diff
@@ -48,20 +48,23 @@ Create containerd service drop-in:
-        {%- set no_proxy = ["localhost", "127.0.0.1"] + networks.values() %}
-        {%- if proxies.no_proxy | default %}
-          {%- do no_proxy.extend(proxies.no_proxy) %}
+        args:
```
Minor: I would have named this variable `containerd_options` or `containerd_args`,  as `args` is not crystal clear.

gdemonet replied:
Done

alexandre-allard commented on salt/metalk8s/container-engine/containerd/installed.sls:
```diff
@@ -84,4 +87,7 @@ Configure registry IP in containerd conf:
-          endpoint = ["http://{{ registry_ip }}:{{ registry_port }}"]
+        endpoint = ["http://{{ registry_ip }}:{{ registry_port }}"]
+
+        [debug]
+        level = "{{ 'debug' if metalk8s.debug else 'info' }}"
```
What's the purpose of using both the containerd option `--log-level` and this ? I think we should stick to only one way of doing this, and putting the option in this file seems the better option to me.

gdemonet replied:
So:
- `--log-level` seems to adjust the main daemon's logging, which ends up in journald
- the `[debug]` config section configures the debug socket, at `/run/containerd/debug.sock` by default (I didn't try listening on this socket so I don't know what gets sent in there :sweat_smile:)

alexandre-allard replied:
Oh ok, I see the difference, I thought it was the same thing, I've checked the man (containerd-config.toml) and effectively it is not related.
Still, that's weird that there is no option to change daemon log level within this configuration file.

alexandre-allard commented on salt/metalk8s/salt/master/configured.sls:
```diff
@@ -12,6 +13,7 @@ Configure salt master:
     - backup: false
     - template: jinja
     - defaults:
+        debug: {{ metalk8s.debug | tojson }}
```
is the `| tojson` filter really necessary ?

gdemonet replied:
I think I had issues with this value being interpreted as a string, but I'll see if I can remove it :)

gdemonet replied:
Removed!

alexandre-allard commented on eve/main.yml:
```diff
@@ -243,6 +246,7 @@ models:
       name: Create bootstrap configuration file
       env:
         PRODUCT_VERSION: "%(prop:metalk8s_version)s"
+        DEBUG: "%(prop:metalk8s_debug:-false)s"
```
Could be cool if all `env` could inherit from a `base_env` which sets this `DEBUG` variable, but it means changing much more stuff, so... nevermind. :)

gdemonet replied:
Yeaaah... Could be cool to have a better test manager than YAML.

alexandre-allard commented on eve/workers/openstack-multiple-nodes/terraform/nodes.tf:
```diff
@@ -92,7 +92,10 @@ resource "openstack_compute_instance_v2" "bootstrap" {
   provisioner "remote-exec" {
-    inline = ["sudo bash scripts/bootstrap-config.sh"]
+    inline = [
+      "sudo chmod +x scripts/bootstrap-config.sh",
```
Can't we simply make this script executable directly on the repository, so we don't need to `chmod` it each time.
I don't know if the file provisioner keep the mode of a file, but I think it is.

gdemonet replied:
Done

alexandre-allard commented on eve/workers/openstack-single-node-rhel/terraform/bootstrap.tf:
```diff
@@ -41,7 +41,8 @@ resource "openstack_compute_instance_v2" "bootstrap" {
   provisioner "remote-exec" {
     inline = [
-      "sudo bash scripts/bootstrap-config.sh",
+      "sudo chmod +x scripts/bootstrap-config.sh",
```
ditto

gdemonet replied:
Done

alexandre-allard reviewed (commented):
Details

gdemonet committed:
vagrant: Add optional debug in BootstrapConfig

alexandre-allard commented on salt/metalk8s/container-engine/containerd/files/50-metalk8s.conf.j2:
```diff
@@ -1,4 +1,11 @@
+# Allow passing arguments to containerd daemon
+ExecStart=
+{%- if containerd_args %}
+ExecStart=/usr/bin/containerd "{{ args | join('" "') }}"
```
Don't forget to also change the name here:
```suggestion
ExecStart=/usr/bin/containerd "{{ containerd_args | join('" "') }}"
```

gdemonet replied:
:facepalm:

bert-e commented:
# History mismatch



Merge commit #5f97664e1e85e3f507daf00cf83f7f811624eff7 on the integration branch
`w/2.6/improvement/add-debug-pillar` is merging a branch which is neither the current
branch `improvement/add-debug-pillar` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/add-debug-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-debug-pillar` into
`development/2.5`
* [w/2.6/improvement/add-debug-pillar](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/add-debug-pillar) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

alexandre-allard commented on eve/main.yml:
```diff
@@ -1103,7 +1109,9 @@ stages:
-          env: *_env_version_prev
+          env:
+            <<: *_env_version_prev
+            DEBUG: "%(prop:metalk8s_debug)s"
```
No default value to fallback on, as you did above:
```suggestion
            DEBUG: "%(prop:metalk8s_debug:-false)s"
```

gdemonet replied:
Done

alexandre-allard commented on eve/main.yml:
```diff
@@ -1408,7 +1416,9 @@ stages:
-          env: *_env_promoted_version
+          env:
+            <<: *_env_promoted_version
+            DEBUG: "%(prop:metalk8s_debug)s"
```
ditto
```suggestion
            DEBUG: "%(prop:metalk8s_debug:-false)s"
```

gdemonet replied:
Done

bert-e commented:
# History mismatch



Merge commit #384ab141bd59d8e4578cbbd5ed30fe9ac5989dc2 on the integration branch
`w/2.6/improvement/add-debug-pillar` is merging a branch which is neither the current
branch `improvement/add-debug-pillar` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/add-debug-pillar` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/add-debug-pillar` into
`development/2.5`
* [w/2.6/improvement/add-debug-pillar](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/add-debug-pillar) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

alexandre-allard reviewed (approved):
LGTM, brace yourselves, debugging logs are coming.

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/91ca31c116132e0e3240f425bed27477adc9f9d7) did not succeed in branch w/2.6/improvement/add-debug-pillar.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye gdemonet.
